egs/librispeech/asr/simple_v1/train.py

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env python3
+
+import logging
+import os
+import sys
+import warnings
+from concurrent.futures import ProcessPoolExecutor
+from pathlib import Path
+
+import numpy as np
+import torch
+import torch.optim as optim
+from torch.nn.utils import clip_grad_value_
+import torchaudio
+import torchaudio.models
+import k2
+
+from lhotse import CutSet, Fbank, LilcomFilesWriter, WavAugmenter
+from lhotse.dataset import SpeechRecognitionDataset
+from lhotse.dataset.speech_recognition import K2DataLoader, K2SpeechRecognitionDataset, \
+    K2SpeechRecognitionIterableDataset, concat_cuts
+from lhotse.recipes.librispeech import download_and_untar, prepare_librispeech, dataset_parts_full
+
+from common import load_checkpoint
+from common import save_checkpoint
+from common import save_training_info
+from common import setup_logger
+from wav2letter import Wav2Letter
+
+
+def create_decoding_graph(texts, graph, symbols):
+    fsas = []
+    for text in texts:
+        filter_text = [
+            i if i in symbols._sym2id else '<UNK>' for i in text.split(' ')
+        ]
+        word_ids = [symbols.get(i) for i in filter_text]
+        fsa = k2.linear_fsa(word_ids)
+        fsa = k2.arc_sort(fsa)
+        decoding_graph = k2.intersect(fsa, graph).invert_()
+        decoding_graph = k2.add_epsilon_self_loops(decoding_graph)
+        fsas.append(decoding_graph)
+    return k2.create_fsa_vec(fsas)
+
+
+def get_objf(batch, model, device, graph, symbols, training, optimizer=None):
+    feature = batch['features']
+    supervisions = batch['supervisions']
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'], supervisions['start_frame'],
+         supervisions['num_frames']), 1).to(torch.int32)
+    texts = supervisions['text']
+    assert feature.ndim == 3
+    #print(feature.shape)
+    #print(supervision_segments[:, 1] + supervision_segments[:, 2])
+
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+    feature = feature.to(device)
+    if training:
+        nnet_output = model(feature)
+    else:
+        with torch.no_grad():
+            nnet_output = model(feature)
+
+    # nnet_output is [N, C, T]
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    # TODO(haowen): create decoding graph at the beginning of training
+    decoding_graph = create_decoding_graph(texts, graph, symbols)
+    decoding_graph.to_(device)
+    decoding_graph.scores.requires_grad_(False)
+    #print(nnet_output.shape)
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+    #dense_fsa_vec.scores.requires_grad_(True)
```

Comment:
`dense_fsa_vec.scores` is a reference to `nnet_output`, so I think this statement
can be removed.

Reply:
sure, this is just for debugging purpose back to that time.

Reply:
Removed